src/cloud-api-adaptor/docker/prereqs.sh

```diff
@@ -0,0 +1,336 @@
+# Following are the packages that are installed:
+# go=https://go.dev/dl/go1.22.4.linux-amd64.tar.gz
+# yq=https://github.com/mikefarah/yq/releases/download/v4.44.2/yq_linux_amd64
+# kubectl=https://storage.googleapis.com/kubernetes-release/release/v1.27.11/bin/linux/amd64/kubectl
+# kind=https://kind.sigs.k8s.io/dl/v0.23.0/kind-linux-amd64
+
+install_binary_packages() {
+    # Define the required binary packages
+    REQUIRED_BINARY_PACKAGES=(
+        "yq=https://github.com/mikefarah/yq/releases/download/v4.44.2/yq_linux_amd64"
+        "kubectl=https://storage.googleapis.com/kubernetes-release/release/v1.27.11/bin/linux/amd64/kubectl"
+        "kind=https://kind.sigs.k8s.io/dl/v0.23.0/kind-linux-amd64"
+    )
+
+    # Specify the installation directory
+    local install_dir="/usr/local/bin"
+
+    # Install the required binary packages
+    for package_info in "${REQUIRED_BINARY_PACKAGES[@]}"; do
+        IFS='=' read -r package_name package_url <<<"${package_info}"
+        download_path="/tmp/${package_name}"
+
+        if [[ -x "${install_dir}/${package_name}" ]]; then
+            echo "Package ${package_name} is already installed. Skipping."
+            continue
+        else
+            echo "Downloading ${package_name}..."
+            # Update a hidden file to record what was installed. This is useful for uninstallation
+            echo "${package_name}" >>"${bin_packages_file}"
+            curl -sSL "${package_url}" -o "${download_path}" ||
+                error_exit "Failed to download ${package_name}"
+
+            echo "Extracting ${package_name}..."
+            if [[ "${package_url}" == *.tar.gz ]]; then
+                sudo tar -xf "${download_path}" -C "${install_dir}" ||
+                    error_exit "Failed to extract ${package_name}"
+            # If not a tar.gz file, then it is a binary file
+            else
+                echo "${package_name} is binary file."
+                sudo mv "${download_path}" "${install_dir}/${package_name}" ||
+                    error_exit "Failed to move ${package_name} to ${install_dir}"
+            fi
+
+            echo "Marking  ${install_dir}/${package_name} executable"
+            sudo chmod +x "${install_dir}/${package_name}" ||
+                error_exit "Failed to mark ${package_name} executable"
+
+            echo "Cleaning up..."
+            rm -f "${download_path}"
+        fi
+    done
+
+    echo "All binary packages installed successfully."
+
+}
+
+# function to install golang
+install_golang() {
+
+    # Check if Go is already installed
+    if [[ -x "$(command -v go)" ]]; then
+        echo "Go is already installed"
+    else
+        # Install Go
+        echo "Installing Go"
+        curl -fsSL https://go.dev/dl/go1.22.4.linux-amd64.tar.gz -o go.tar.gz || error_exit "Failed to download Go"
+        sudo tar -C /usr/local -xzf go.tar.gz || error_exit "Failed to extract Go"
+        touch "$go_file"
+        rm -f go.tar.gz
+    fi
+
+}
```

Comment:
Could we refactor out and re-use https://github.com/confidential-containers/cloud-api-adaptor/blob/526b216ffff640deb14283fbed4536d66ea3e476/src/cloud-api-adaptor/libvirt/config_libvirt.sh#L20-L45, as that automatically installs our supported version of golang for versions.yaml?

Reply:
Can refactoring be a future item?
Reason being the config_libvirt.sh currently aligned (naming, intent etc) with libvirt provider. It does provide method to install golang and kubectl that can be used. But it provides no method for uninstall. Also golang used is just for the test execution code, and not for the core provider code. So we can use latest golang without the risk of any incompatibility with the core code being tested.

I can also remove the prereqs.sh for now as this is more of a helper and one can set up their environment whichever way they want. Later on I can introduce additional library code for env setup that can be used for different providers.
So to summarise the options
1. Keep the prereqs.sh code as-is
2. Remove the prereqs.sh code for now and let users set the environment in their own preferred way.

Reply:
How about 1a - tweak the pre-reqs.sh for now to use the same logic as config_libvirt.sh to get the version from versions.yaml and then we refactor later?